### PR TITLE
[ENG-4302] Add Terraform CI shared workflow

### PR DESCRIPTION
**Pull request by bateller**
JIRA Ticket: https://tradeswell.atlassian.net/browse/ENG-4302

~~Not sure how exactly to test this. I attempted [act](https://github.com/nektos/act) but it didn't handle the workflow calls correctly~~ Edit: Tested using https://github.com/Tradeswell/devops-github-actions-test-repo

### CODE REVIEW

**Comment by bateller** [comment]
Test file
```
name: Terraform CI
on:
  push:
    branches: [ main ]
    paths-ignore:
      - '**/README.md'
      - '.gitignore'
      - 'docs/**'

  pull_request:
    branches: [ main ]
    paths-ignore:
      - '**/README.md'
      - '.gitignore'
      - 'docs/**'

jobs:
  deploy:
    # Only deploy off the main branch
    if: github.ref == 'refs/heads/main'
    uses: Tradeswell/shared-workflows/.github/workflows/terraform_ci.yml@main
    with:
      workspaces: 'prod'
      working_dir: './'
    secrets:
      TF_API_TOKEN: ${{ secrets.TF_API_TOKEN }}
      GH_SSH_PRIVATE_KEY: ${{ secrets.GH_SSH_PRIVATE_KEY }}
```

File: .github/workflows/terraform_ci.yml [hunk 1]
```diff
@@ -0,0 +1,66 @@
+name: 'Deploy'
+on:
+  workflow_call:
+    inputs:
+      workspaces:
+        type: list
```
Comment (kherkeladze) [inline .github/workflows/terraform_ci.yml]: Can't find `list` type support in [docs](https://docs.github.com/en/enterprise-cloud@latest/actions/using-workflows/workflow-syntax-for-github-actions#onworkflow_callinputsinput_idtype):

> The value of this parameter is a string specifying the data type of the input. This must be one of: boolean, number, or string.

One possible solution I found: [reusable-workflow-with-strategy-matrix](https://github.community/t/reusable-workflow-with-strategy-matrix/205676/2)

Comment (bateller) [reply .github/workflows/terraform_ci.yml]: I tried `string` (you can see how a single entry is used on line 43: https://github.com/Tradeswell/shared-workflows/pull/2/files#diff-98941fce34014d35c8a4b01fb52dbd40ae583ff89bb5ebf6b3f1484dfd233bc8R43) with no luck too :(

Comment (bateller) [reply .github/workflows/terraform_ci.yml]: Doesn't look like you can have matrices inside steps, so I duplicated the jobs for each env/dir. It isn't as streamlined as I'd like, but it works:

```
name: Terraform CI
on:
  push:
    branches: 
      - main
    paths-ignore:
      - '**/README.md'
      - '.gitignore'
      - 'docs/**'

  pull_request:
    branches: 
      - main
    paths-ignore:
      - '**/README.md'
      - '.gitignore'
      - 'docs/**'

jobs:
  prod:
    uses: Tradeswell/shared-workflows/.github/workflows/terraform_ci.yml@ENG-3996
    with:
      workspaces: 'prod'
      working_dir: './prod'
    secrets:
      TF_API_TOKEN: ${{ secrets.TF_API_TOKEN }}
      GH_SSH_PRIVATE_KEY: ${{ secrets.GH_SSH_PRIVATE_KEY }}
  qa:
    uses: Tradeswell/shared-workflows/.github/workflows/terraform_ci.yml@ENG-3996
    with:
      workspaces: 'qa'
      working_dir: './qa'
    secrets:
      TF_API_TOKEN: ${{ secrets.TF_API_TOKEN }}
      GH_SSH_PRIVATE_KEY: ${{ secrets.GH_SSH_PRIVATE_KEY }}
  dev:
    uses: Tradeswell/shared-workflows/.github/workflows/terraform_ci.yml@ENG-3996
    with:
      workspaces: 'dev'
      working_dir: './dev'
    secrets:
      TF_API_TOKEN: ${{ secrets.TF_API_TOKEN }}
      GH_SSH_PRIVATE_KEY: ${{ secrets.GH_SSH_PRIVATE_KEY }}
```

**Review by kherkeladze — DISMISSED** [review]

**Review by donthorp — APPROVED** [review]
Looks like it performs the required tasks.

**Comment by bateller** [comment]
Latest successful test can be found here: https://github.com/Tradeswell/devops-github-actions-test-repo/actions/runs/2436318068

**Review by kherkeladze — APPROVED** [review]

**Pull request merged**